### ISSUE TITLE
ServiceFabricComposeDeploy: A communication error caused the operation to fail.

### ISSUE THREAD

**Issue by CodedBeard** [Microsoft/vsts-tasks#5161]
I've been using 2 test clusters with the preview release for service fabric 5.7, which have been working fine, but as 5.7 has been released now, I decided to go ahead and build a new cluster, using the same ARM template as before, just updating the version from preview to release.

This appeared to work fine, the cluster came online and I was able to browse the cluster explorer, and connect to the cluster via PowerShell. However, I then attempted to release a docker-compose application from VSTS onto the new cluster, which fails when calling "New-ServiceFabricComposeApplication" with the error "A communication error caused the operation to fail.". I then went back to PowerShell and tried running the same commands myself, and the compose application gets created just fine without an error.

I have double checked the release configuration, and ensured all the settings are the same as the previous clusters, with the only difference being the host name for the client endpoint. I can see in the release logs that "Successfully connected to cluster." is reported before attempting to create the application.

I have now re-run the deployment with "system.debug=true", which gave the following stack trace:
"System.Fabric.FabricException: A communication error caused the operation to fail. ---> System.Runtime.InteropServices.COMException: Exception from HRESULT: 0x80071BBC
System.Fabric.Interop.NativeClient.IInternalFabricApplicationManagementClient.EndCreateComposeApplication(IFabricAsyncOperationContext context)
at System.Fabric.Interop.Utility.<>c__DisplayClass20_0.<WrapNativeAsyncInvoke>b__0(IFabricAsyncOperationContext context)
at System.Fabric.Interop.AsyncCallOutAdapter2`1.Finish(IFabricAsyncOperationContext context, Boolean expectedCompletedSynchronously)
--- End of inner exception stack trace ---"

Any ideas what could be causing this exception from VSTS based release only?

**Comment by bishal-pdMSFT** [comment]
@CodedBeard 
1. is this consistent or a transient connection error problem?
2. what agent pool are you using? is it private or hosted or hostedVS2017?
3. Please attach a gist containing task log when "system.debug=true"

**Comment by CodedBeard** [comment]
@bishal-pdMSFT 

1. Consistent, It hasn't worked once with the new cluster. I even built another new cluster to make sure it wasn't some random issue with the first cluster.
2. HostedVS2017

Will get the log output cleaned and uploaded shortly.

Another thing I noticed yesterday, If I leave the application deployed via manual PowerShell, and then run the vsts release, it correctly detects the application is already deployed, removes it, and then fails when calling "New-ServiceFabricComposeApplication" again.

**Comment by CodedBeard** [comment]
@bishal-pdMSFT 

Here's the output for the release targeting the 5.7 "preview" cluster, which works as expected: https://gist.github.com/CodedBeard/84cfa1651792f49bab124d1caa7cd9f0

Here's the output for the first 5.7 "release" cluster, which always fails from VSTS, but works fine from PowerShell:
https://gist.github.com/CodedBeard/a5208b278faee2d817e644bcc83915b9

Finally, here's the output from the 2nd 5.7 "release" cluster, when the deployment from PowerShell is left in place before running a release from VSTS, which shows the application being detected, removed, and then failing on creation again:
https://gist.github.com/CodedBeard/2a44c8316cb8366eff519b9540c7dc79

**Comment by bishal-pdMSFT** [comment]
@CodedBeard 
HostedVS2017 agent is using 2.7 **preview** SDK version. There could be compatibility issue between 2.7 preview SDK and 5.7 release runtime. Are you using 2.7 **release** SDK when running commands through powershell directly?
@MikkelHegn, are you aware of any compat issue?

**Comment by CodedBeard** [comment]
@bishal-pdMSFT yea, the machine I'm running the commands from is using the 2.7 release SDK

**Comment by bishal-pdMSFT** [comment]
@CodedBeard Would you be able to use a private agent with 2.7 release SDK to un-block yourself. Are you able to deploy to 2.7 preview cluster using your machine with powershell commands?

**Comment by CodedBeard** [comment]
@bishal-pdMSFT will give the private agent a go tomorrow.

**Comment by CodedBeard** [comment]
@bishal-pdMSFT I've just confirmed it works fine when using a private agent with the 2.7 release SDK, so looks like there is an issue with the HostedVS2017 agent's version

**Comment by bishal-pdMSFT** [comment]
Thanks @CodedBeard 
@MikkelHegn do you see any issue if we update HostedVS2017 image's SF SDK to 2.7 version?

**Comment by MikkelHegn** [comment]
No we should go ahead and do that.

**Comment by FelschR** [comment]
@bishal-pdMSFT @MikkelHegn 
It seems the VSTS images weren't updated with the new SDK yet?

**Comment by bishal-pdMSFT** [comment]
Assigning this issue to @chrisrpatterson as he own updating hosted agent images

**Comment by the-daniel-rothig** [comment]
The issue persists - any guess when it might get resolved?

**Comment by bishal-pdMSFT** [comment]
The SF SDK version has been updated to 2.7 on hostedVS2017 pool. Do let me know if you are still facing this problem.

**Comment by CodedBeard** [comment]
thanks @bishal-pdMSFT, to be honest, I've stopped using the hosted agents completely, As service fabric 6.0 has already been released before 5.7 got deployed, and we can't be waiting 6 weeks between SF releases before we can actually use the release pipeline. We've also hugely benefited from using the private build agents due to caching, as we are migrating old systems into containers, which being based on server core are quite large. This was resulting in hosted builds taking almost 30 mins, most of which was just re-downloading the server core image every time, using private the builds went from 24 mins, to just over a min, so we will be sticking with this until the hosted build agent does a better job of caching container images.
